drivers/SmartThings/zwave-garage-door-opener/src/ecolink-zw-gdo/init.lua

```diff
@@ -0,0 +1,309 @@
+
+local CONTACTSENSOR_BATTERY_LEVEL_NORMAL = 100
+local CONTACTSENSOR_BATTERY_LEVEL_LOW = 1
+
+local GDO_CONFIG_PARAM_NO_UNATTENDED_WAIT = 1
+local GDO_CONFIG_PARAM_NO_ACTIVATION_TIME = 2
+local GDO_CONFIG_PARAM_NO_OPEN_TIMEOUT = 3
+local GDO_CONFIG_PARAM_NO_CLOSE_TIMEOUT = 4
+local GDO_CONFIG_PARAM_NO_SHAKE_SENSE = 5
+local GDO_CONFIG_PARAM_NO_APP_RETRY = 6
+
+--- Determine whether the passed device is an Ecolink garage door operator
+---
+--- @param driver Driver driver instance
+--- @param device Device device isntance
+--- @return boolean true if the device proper, else false
+local function can_handle_ecolink_garage_door(opts, driver, device, ...)
+  for _, fingerprint in ipairs(ECOLINK_GARAGE_DOOR_FINGERPRINTS) do
+    if
+    device:id_match(fingerprint.manufacturerId, fingerprint.productType, fingerprint.productId)
+    then
+      return true
+    end
+  end
+  return false
+end
+
+local function component_to_endpoint(device, component_id)
+  if (CONTACTSENSOR_ENDPOINT_NAME == component_id)  then
+    --contactSensor is 2
+    return CONTACTSENSOR_ENDPOINT_NUMBER
+  end
+  -- main endpoint is garage door
+  return GDO_ENDPOINT_NUMBER
+end
+
+local function endpoint_to_component(device, ep)
+  if ( CONTACTSENSOR_ENDPOINT_NUMBER == ep ) then
+    return CONTACTSENSOR_ENDPOINT_NAME
+  end
+  return GDO_ENDPOINT_NAME
+end
+
+--- Handle Device Instantiated Event
+---
+--- @param driver st.zwave.Driver
+--- @param device st.zwave.Device
+local function device_instantiated(driver, device)
+  log.info_with({hub_logs=true}, "device init")
+  device:set_component_to_endpoint_fn(component_to_endpoint)
+  device:set_endpoint_to_component_fn(endpoint_to_component)
+  device.thread:call_with_delay(constants.DEFAULT_GET_STATUS_DELAY, function(d)
+    device:send(BarrierOperator:Get({}))
+  end)
+  device.thread:call_with_delay(constants.DEFAULT_GET_STATUS_DELAY*2, function(d)
+    device:send(SensorMultilevel:Get({}))
+  end)
+  device.thread:call_with_delay(constants.DEFAULT_GET_STATUS_DELAY*3, function(d)
+    device:send(Configuration:BulkGetV2({parameter_offset = 1, number_of_parameters = 6}) )
+  end)
+end
+
+--- Handle Device Added Event
+---
+--- @param driver st.zwave.Driver
+--- @param device st.zwave.Device
+local function device_added(driver, device)
+  device:send(BarrierOperator:Get({}))
+  -- Reset contact sensor battery level... This should be pollable on the GDO side.
+  device:emit_event_for_endpoint(CONTACTSENSOR_ENDPOINT_NUMBER,
+  capabilities.battery.battery(CONTACTSENSOR_BATTERY_LEVEL_NORMAL)
+  )
+  -- Reset contact sensor fields
+  device:emit_event_for_endpoint(CONTACTSENSOR_ENDPOINT_NUMBER,
+    capabilities.tamperAlert.tamper.clear()
+    )
+  device:emit_event_for_endpoint(CONTACTSENSOR_ENDPOINT_NUMBER,
+    capabilities.contactSensor.contact.closed()
+    )
+  -- Init barrier door state
+  device:emit_event_for_endpoint(GDO_ENDPOINT_NUMBER,
+    capabilities.doorControl.door.closed()
+    )
```

Comment:
nit: the spacing here is a bit weird/inconsistent

Reply:
I changed it to make the spacing a bit more consistent and as clear as possible here.